### PR TITLE
Keep an eye on Python 3.7

### PR DESCRIPTION
**Pull request by jacg**
Added Python 3.7 as an allowed failure to the Travis build matrix. Like this we will know when all our dependencies are available on conda in Python 3.7.

I have also taken this opportunity to:

+ Update the pinned versions of our dependencies to the most recent ones
+ Update conda to the most recent version, and change the acivation command to the one which has been available and recommended since 4.4 (#455).

### CODE REVIEW

**Comment by jacg** [comment]
I have a feeling that @jjgomezcadenas will like [dataclasses](https://www.python.org/dev/peps/pep-0557/) which have been added in Python 3.7.

**Comment by jacg** [comment]
@jjgomezcadenas `test_spe_to_adc` is the only remaining test to fail after the package upgrades. Can you have a look to see if you can spot what has changed?

**Comment by jacg** [comment]
@jjgomezcadenas The culprit seems to be `scipy` which has changed from version 0.19 to 1.1, so an API change is to be expected.

**Comment by jjgomezcadenas** [comment]
@jacg to be sure: an API change from our side? Sorry about being thick, travelling right now (Korea) and with jet lag.

**Comment by jjgomezcadenas** [comment]
Yes, I like dataclasses very much, they extend the funcionality I am using all the type with named tuples.

**Comment by jacg** [comment]
@jjgomezcadenas No, you can expect an API change in scipy when its version number changes from 0. something to 1.whatever. `test_spe_to_adc` is now failing, but if you roll back scipy from 1.1.0 to 0.19.1 then `test_spe_to_adc` passes.

No point in hanging about in the ancient past (we'll leave that sort of stuff to CERN), so we need to adapt `test_spe_to_adc` to deal with whatever change occurred in scipy between 0.19.1 and 1.1.0.

**Comment by jjgomezcadenas** [comment]
OK, I will look at it as soon as I can.

**Comment by jacg** [comment]
Oh, and before you start looking into this, a word of warning: testing this is a huge PITA, because you need a *different conda environment* for this PR (and anything that is rebased on top of it) while anything that comes before this PR needs the old version of the conda env. Our `manage.sh` utilities cannot (at the moment) distinguish between these two environments, they use the same name for both.

It's not immediately obvious how to deal with it nicely, and to what extent it is worthwhile to build the capability into `manage.sh`. This is a problem we will have to face repeatedly but rarely: every time we want to bump the versions of dependencies.

As an ad-hoc solution, you could try something along the lines of

1. Rename the old env:
    + `conda create -n IC3.6-old --clone IC3.6`
    + `conda remove -n IC3.6 --all`
2. Create a new env for post version-bump work:
    + open new shell
    + `cd $ICDIR`
    + `git checkout python-3.7` (ensure that you have the new versions in `manage.sh`)
    + `source manage.sh make_environment 3.6`

This should give you two conda envs `IC3.6` and `IC3.6-old`, and you'll need to make sure that you are using the appropriate one, depending on whether you are working on a commit which comes before or after the version bump.

**Comment by jacg** [comment]
@jjgomezcadenas #496 should make it easier to investigate the problem with `test_spe_to_adc`, so I suggest you work on it there.

**Comment by jacg** [comment]
The commits in this PR which address dependency evolution, have a more natural place in #496, so **do not merge this** until #496 has been merged or rejected.

**Comment by jacg** [comment]
It seems that Python 3.7 is IC-ready: all our tests pass on Python 3.7.

Do we want to keep Python 3.6 for a while, or shall we switch to 3.7 straight away?_

**Comment by jjgomezcadenas** [comment]
I need to test that 3.7 works in my system, will do it later today or tomorrow (today is Sunday, we are out visiting Chicago)

**Comment by jjgomezcadenas** [comment]
we are using conda 4.5.4, the latest version is 4.5.8. Since we are a it, should we bump conda too?

**Review by jjgomezcadenas — APPROVED** [review]
python 3.7 passes all tests in my machine. As an extra check I have run a complex analysis notebook that performs all kind of fits to the lifetime of the detector and the results are identical in 3.6 and 3.7. Thus, it seems that the upgrade is a success. 

I recommend keeping both versions of python for a short while. The current software in production is very stable and I am reluctant to introduce any changes which are not needed in the prod version, but the dev version of the software can already go to the bumped-package, 3.7 python. Over the next couple of weeks we can launch a test production of kdsts with dev and verify that everything is fine. 

However I would recommend all developers to move to the bumped 3.7 version, so that once we move the dev (3.7) to prod, we don't need to support 3.6 anymore (don't see the point, once we have checked everything is fine and transition has been completed. 

Please, @jacg if you can merge right away (let @carmenromo learn with the PR we have assigned to her)

**Comment by jacg** [comment]
> we are using conda 4.5.4, the latest version is 4.5.8. Since we are a it, should we bump conda too?

Yes, I was thinking along those lines too. Of course, the conda version can't really get switched each time we switch historical environments, there's some scope here for inonsistency. I don't see a way around it, though.

**Comment by jacg** [comment]
> we are using conda 4.5.4, the latest version is 4.5.8. Since we are a it, should we bump conda too?

There is no installer for 4.5.8, the most recent one is 4.5.4, so I suggest we leave this until later.

**Comment by jjgomezcadenas** [comment]
Agreed, let's leave 4.5.4 but something went wrong here...

**Comment by jacg** [comment]
> something went wrong here...

What goes wrong (most often) is that hypothesis sometimes finds counterexamples in `test_inverse_cdf_index_hypothesis_generated`. I get the impression that it finds counterexamples much more frequently now, than it did in the past, but because 'the past' is before #496 (Dependency versioning) it's *extremely* painful to track down and understand exactly what is responsible for this effect (and if it really exists ... the inherent randomness and slow turnaround times really don't help either).

Anyway, I'm just going to keep pressing the *Restart job* button on Travis, until I get a pass, so that it's formally approved by Travis. Then I'll open another PR and try to get to the bottom of this. But I need to decouple that issue from this PR, otherwise I'll never make any progress.

Sometimes if fails because of hypothesis health checks, but that seems to be much rarer, and it happens in completely different tests.

**Pull request merged**
